File: app/src/main/java/org/wikipedia/edit/Edit.java
package org.wikipedia.edit;

import androidx.annotation.Nullable;

import org.wikipedia.dataclient.mwapi.MwPostResponse;

public class Edit extends MwPostResponse {
    @SuppressWarnings("unused,") @Nullable private Result edit;

    @Nullable public Result edit() {
        return edit;
    }

    boolean hasEditResult() {
        return edit != null;
    }

    public class Result {
        @SuppressWarnings("unused") @Nullable private String result;
        @SuppressWarnings("unused") private long newrevid;
        @SuppressWarnings("unused") @Nullable private Captcha captcha;
        @SuppressWarnings("unused") @Nullable private String code;
        @SuppressWarnings("unused") @Nullable private String info;
        @SuppressWarnings("unused") @Nullable private String warning;
        @SuppressWarnings("unused") @Nullable private String spamblacklist;

        @Nullable String status() {
            return result;
        }

        public long newRevId() {
            return newrevid;
        }

        boolean editSucceeded() {
            return "Success".equals(result);
        }

        @Nullable String captchaId() {
            return captcha == null ? null : captcha.id();
        }

        boolean hasEditErrorCode() {
            return code != null;
        }

        boolean hasCaptchaResponse() {
            return captcha != null;
        }

        @Nullable String code() {
            return code;
        }

        @Nullable String info() {
            return info;
        }

        @Nullable String warning() {
            return warning;
        }

        @Nullable String spamblacklist() {
            return spamblacklist;
        }

        boolean hasSpamBlacklistResponse() {
            return spamblacklist != null;
        }
    }

    private static class Captcha {
        @SuppressWarnings("unused") @Nullable private String id;

        @Nullable String id() {
            return id;
        }
    }
}

File: app/src/main/java/org/wikipedia/edit/Edit.kt
package org.wikipedia.edit

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import org.wikipedia.dataclient.mwapi.MwPostResponse

@Serializable
class Edit : MwPostResponse() {

    val edit: Result? = null

    @Serializable
    class Result {
        private val captcha: Captcha? = null
        @SerialName("result")
        val status: String? = null
        @SerialName("newrevid")
        val newRevId: Long = 0
        val code: String? = null
        val info: String? = null
        val warning: String? = null
        val spamblacklist: String? = null

        val editSucceeded get() = "Success" == status
        val captchaId get() = captcha?.id.orEmpty()
        val hasEditErrorCode get() = code != null
        val hasCaptchaResponse get() = captcha != null
        val hasSpamBlacklistResponse get() = spamblacklist != null
    }

    @Serializable
    private class Captcha {
        val id: String? = null
    }
}
